### PR TITLE
feat(fsdrv): implement arduino littlefs driver

### PR DESCRIPTION
**Pull request by PierreRambaud**
### Description of the feature or fix

Currently working in progress, it works but I'm not satisfied with the C++ code.

### Notes
- Update the [Documentation](https://github.com/lvgl/lvgl/tree/master/docs) if needed.
- Add [Examples](https://github.com/lvgl/lvgl/tree/master/examples) if relevant.
- Add [Tests](https://github.com/lvgl/lvgl/blob/master/tests/README.md) if applicable.
- If you added new options to `lv_conf_template.h` run [lv_conf_internal_gen.py](https://github.com/lvgl/lvgl/blob/master/scripts/lv_conf_internal_gen.py) and update [Kconfig](https://github.com/lvgl/lvgl/blob/master/Kconfig).
- Run `scripts/code-format.py` ([astyle](http://astyle.sourceforge.net/install.html) version [v3.4.10](https://github.com/szepeviktor/astyle/releases/tag/v3.4.10) needs to be installed) and follow the [Code Conventions](https://docs.lvgl.io/master/CODING_STYLE.html).
- Mark the Pull request as [Draft](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-stage-of-a-pull-request) while you are working on the first version, and mark is as _Ready_ when it's ready for review.
- When changes were requested, [re-request review](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/requesting-a-pull-request-review) to notify the maintainers.
- Help us to review this Pull Request! Anyone can [approve or request changes](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/approving-a-pull-request-with-required-reviews).

### CODE REVIEW

**Comment by kisvegabor** [comment]
> I'm not satisfied with the C++ code.

I'm okay with that. It's simple and clean. 


Please don't forget to update the docs. :slightly_smiling_face:

File: src/libs/fsdrv/lv_fs_arduino_littlefs.cpp [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+static lv_fs_res_t fs_seek(lv_fs_drv_t * drv, void * file_p, uint32_t pos, lv_fs_whence_t whence);
+static lv_fs_res_t fs_tell(lv_fs_drv_t * drv, void * file_p, uint32_t * pos_p);
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+/**
+ * Register a driver for the LittleFS File System interface
+ */
+void lv_fs_arduino_littlefs_init(void)
```
Comment (FASTSHIFT) [inline src/libs/fsdrv/lv_fs_arduino_littlefs.cpp]: extern "C" void lv_fs_arduino_littlefs_init(void)
{
}

File: src/libs/fsdrv/lv_fs_arduino_littlefs.cpp [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+ **********************/
+static void fs_init(void);
+static void * fs_open(lv_fs_drv_t * drv, const char * path, lv_fs_mode_t mode);
+static lv_fs_res_t fs_close(lv_fs_drv_t * drv, void * file_p);
+static lv_fs_res_t fs_read(lv_fs_drv_t * drv, void * file_p, void * buf, uint32_t btr, uint32_t * br);
+static lv_fs_res_t fs_write(lv_fs_drv_t * drv, void * file_p, const void * buf, uint32_t btw, uint32_t * bw);
+static lv_fs_res_t fs_seek(lv_fs_drv_t * drv, void * file_p, uint32_t pos, lv_fs_whence_t whence);
+static lv_fs_res_t fs_tell(lv_fs_drv_t * drv, void * file_p, uint32_t * pos_p);
+
+#ifdef __cplusplus
```
Comment (FASTSHIFT) [inline src/libs/fsdrv/lv_fs_arduino_littlefs.cpp]: remove

File: src/libs/fsdrv/lv_fs_arduino_littlefs.cpp [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+
+    fs_drv->dir_close_cb = NULL;
+    fs_drv->dir_open_cb = NULL;
+    fs_drv->dir_read_cb = NULL;
+
+    lv_fs_drv_register(fs_drv);
+}
+
+#ifdef __cplusplus
+}
```
Comment (FASTSHIFT) [inline src/libs/fsdrv/lv_fs_arduino_littlefs.cpp]: remove

**Review by FASTSHIFT — CHANGES_REQUESTED** [review]

**Review by FASTSHIFT — DISMISSED** [review]

**Comment by PierreRambaud** [comment]
> > I'm not satisfied with the C++ code.
> 
> I'm okay with that. It's simple and clean.
> 
> Please don't forget to update the docs. 🙂

It should be better right now

File: docs/libs/arduino-littlefs.rst [hunk 4]
```diff
@@ -0,0 +1,22 @@
+.. _arduino_littlefs:
+
+==============
+Arduino littlefs
+==============
```
Comment (kisvegabor) [inline docs/libs/arduino-littlefs.rst]: The length of the decoration and the length of the text should match.

```suggestion
================
Arduino littlefs
================
```

File: docs/libs/arduino-littlefs.rst [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+
+LittleFS is a little fail-safe filesystem designed for microcontrollers and integrated in the Arduino framework 
+when used with ESP32 and ESP8266.
```
Comment (kisvegabor) [inline docs/libs/arduino-littlefs.rst:8]: Isn't there a generic Arduino-LittleFS implementation?

**Review by kisvegabor — COMMENTED** [review]
Thank you! :rocket:

File: docs/libs/arduino-littlefs.rst (hunk 5 above)
Comment (PierreRambaud) [reply docs/libs/arduino-littlefs.rst:8]: I didn't find anything too generic for LittleFS, each time I dig into it, it's related to ESP, even if I use platformio and require `FS` + `LittleFS` it targets ESP systems.

Comment (kisvegabor) [reply docs/libs/arduino-littlefs.rst:8]: I see. In this case we should indicate in the name of driver that it's related to ESP only.

Comment (PierreRambaud) [reply docs/libs/arduino-littlefs.rst:8]: wdyt of arduino_esp_littlefs?

Comment (kisvegabor) [reply docs/libs/arduino-littlefs.rst:8]: I buy it :slightly_smiling_face:

**Review by kisvegabor — DISMISSED** [review]
Looks good to me! Thank you! 

The next step will be the SD card library?

**Review by XuNeo — COMMENTED** [review]
I'm not sure how to use the driver, but shouldn't the `lvgl/env_support/cmake/custom.cmake` be updated to include the driver to build?

File: docs/libs/fs.rst [hunk 6]
```diff
@@ -15,6 +15,7 @@ LVG has built in support for:
 - WIN32 (Windows using Win32 API function .e.g ``CreateFileA``, ``ReadFile``)
 - MEMFS (read a file from a memory buffer)
 - LITTLEFS (a little fail-safe filesystem designed for microcontrollers)
+- Arduino LITTLEFS (a little fail-safe filesystem designed for Arduino ESP)
```
Comment (PGNetHun) [inline docs/libs/fs.rst:18]: If this Arduino LittleFS is really ESP dependent, then name should reflect that also.
```suggestion
- Arduino ESP LITTLEFS (a little fail-safe filesystem designed for Arduino ESP)
```

**Review by PGNetHun — CHANGES_REQUESTED** [review]
The code change is basically OK, 
but I'm not sure that we really should add this driver, too.

So basically there are 2 problems with this PR: 
1. "hard-coding" platform-specific 3rd party libraries (even if it's configurable in `lv_conf.h`)
2. dependency on 3 libraries:
    - Arduino
    - ESP (Esperiff library)
    - LittleFS

**Too many dependencies.**

Why not having the FS wrapper in the user's project instead of hard-coding this heavily-depending Arduino-ESP-LittleFS driver to LVGL?

As I remember apps can register their own FS drivers using LVGL API, so it is possible to register any FS handler into LVGL from app: Arduino ESP LittleFS or Arduino SPIFFS, etc....

**Comment by PierreRambaud** [comment]
> The code change is basically OK, but I'm not sure that we really should add this to LVGL.
> 
> So basically there are 2 problems with this PR:
> 
> 1. "hard-coding" platform-specific 3rd party libraries (even if it's configurable in `lv_conf.h`)
> 2. dependency on 3 libraries:
>    
>    * Arduino
>    * ESP (Esperiff library)
>    * LittleFS
> 
> **Too many dependencies.**
> 
> Why not having the FS wrapper in the user's project instead of hard-coding this heavily-depending Arduino-ESP-LittleFS driver to LVGL?
> 
> As I remember apps can register their own FS drivers using LVGL API, so it is possible to register any FS handler into LVGL from app: Arduino ESP LittleFS or Arduino SPIFFS, etc....

In fact, there are only two dependencies, LittleFS is in the Core of the ESP source. But I understand it can be too much for LVGL, depending on what kind of driver it accepts.
I'm not against closing this PR if you think it's not relevant

**Comment by kisvegabor** [comment]
I think this PR (and the planned similar one for the Arduino SD card library) is useful simply because seemingly people are struggling with creating their own FS wrappers. We have a lot of ESP and Arduino users, so it makes sense to have a ready to use solution for them.

**Comment by PGNetHun** [comment]
> I think this PR (and the planned similar one for the Arduino SD card library) is useful simply because seemingly people are struggling with creating their own FS wrappers. We have a lot of ESP and Arduino users, so it makes sense to have a ready to use solution for them.

Well, it makes sense. 

The other option would be to add those FS wrappers to [Arduino examples folder](https://github.com/lvgl/lvgl/tree/master/examples/arduino/LVGL_Arduino) and mention those in [Arduino docs](https://docs.lvgl.io/master/integration/framework/arduino.html), so people can copy-paste the different FS wrappers to their project folder.

**Comment by PierreRambaud** [comment]
> > I think this PR (and the planned similar one for the Arduino SD card library) is useful simply because seemingly people are struggling with creating their own FS wrappers. We have a lot of ESP and Arduino users, so it makes sense to have a ready to use solution for them.
> 
> Well, it makes sense.
> 
> The other option would be to add those FS wrappers to [Arduino examples folder](https://github.com/lvgl/lvgl/tree/master/examples/arduino/LVGL_Arduino) and mention those in [Arduino docs](https://docs.lvgl.io/master/integration/framework/arduino.html), so people can copy-paste those to their project folders.

Tell me what you decide, whatever the decision, I'll be glad to update the PR and help the project.

**Comment by kisvegabor** [comment]
> The other option would be to add those FS wrappers to [Arduino examples folder](https://github.com/lvgl/lvgl/tree/master/examples/arduino/LVGL_Arduino) and mention those in [Arduino docs](https://docs.lvgl.io/master/integration/framework/arduino.html), so people can copy-paste those to their project folders.

In case of Arduino the copy-pasting is not the best way becasue:
1. The Arduino build system is very limited. You can copy files only to your project root folder.
2. I might sound simple to copy the files from here to there, but according to my experience it's not trivial for everyone
3. Once the file is copy-pasted it won't be updated if LVGL is updated.

So I vote for the current solution.

**Comment by XuNeo** [comment]
Does it make sense to change `src/libs/fsdrv/lv_fs_arduino_esp_littlefs.cpp` to a C file instead? LVGL claims to be a C library, the this use case is CPP for sure.

File: Kconfig [hunk 7]
```diff
@@ -1071,6 +1071,13 @@ menu "LVGL configuration"
 			default 0
 			depends on LV_USE_FS_LITTLEFS
 
+		config LV_USE_FS_ARDUINO_ESP_LITTLEFS
+			bool "File system on top of littlefs API"
```
Comment (PGNetHun) [inline Kconfig]:
```suggestion
			bool "File system on top of Arduino ESP littlefs API"
```

File: src/libs/fsdrv/lv_fs_arduino_esp_littlefs.cpp [hunk 8]
```diff
@@ -0,0 +1,181 @@
+#include "../../../lvgl.h"
+#if LV_USE_FS_ARDUINO_ESP_LITTLEFS
+
+#include "../../core/lv_global.h"
+#include "LittleFS.h"
+
+typedef struct ArduinoLittleFile {
```
Comment (PGNetHun) [inline src/libs/fsdrv/lv_fs_arduino_esp_littlefs.cpp]:
```suggestion
typedef struct ArduinoEspLittleFile {
```

File: src/libs/fsdrv/lv_fs_arduino_esp_littlefs.cpp [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+    File file;
+} ArduinoLittleFile;
```
Comment (PGNetHun) [inline src/libs/fsdrv/lv_fs_arduino_esp_littlefs.cpp]:
```suggestion
} ArduinoEspLittleFile;
```

File: src/libs/fsdrv/lv_fs_arduino_esp_littlefs.cpp
Comment (PGNetHun) [inline src/libs/fsdrv/lv_fs_arduino_esp_littlefs.cpp:1]: Is it possible to "convert" this CPP to C ?

**Review by PGNetHun — CHANGES_REQUESTED** [review]
Only some minor changes.

File: src/libs/fsdrv/lv_fs_arduino_esp_littlefs.cpp
Comment (PierreRambaud) [reply src/libs/fsdrv/lv_fs_arduino_esp_littlefs.cpp:1]: If i convert the file in C only I receive:
```In file included from /home/got/.platformio/packages/framework-arduinoespressif32/libraries/LittleFS/src/LittleFS.h:17,
                 from .pio/libdeps/m5stack-core2/lvgl/src/libs/fsdrv/lv_fs_arduino_esp_littlefs.c:5:
/home/got/.platformio/packages/framework-arduinoespressif32/libraries/FS/src/FS.h:24:10: fatal error: memory: No such file or directory
 #include <memory>
          ^~~~~~~~
compilation terminated.
*** [.pio/build/m5stack-core2/libf04/lvgl/libs/fsdrv/lv_fs_arduino_esp_littlefs.c.o] Error 1
```


Like thorvg, I don't think it's a problem to have cpp files if they are not directly used :thinking:

**Comment by PierreRambaud** [comment]
I failed a force push and GitHub closed this PR, I had to recreate a new one :( #5905
Sorry for this mistake

**Pull request closed without merging**